src/hotspot/cpu/x86/stubGenerator_x86_64_tanh.cpp

```diff
@@ -0,0 +1,502 @@
+  __ subl(rcx, 16304);
+  __ cmpl(rcx, 144);
+  __ jcc(Assembler::aboveEqual, L_2TAG_PACKET_0_0_1);
+  __ subsd(xmm4, xmm3);
+  __ mulsd(xmm3, xmm1);
+  __ mulsd(xmm2, xmm5);
+  __ cvtsd2siq(rax, xmm3);
+  __ movq(xmm7, xmm3);
+  __ addsd(xmm3, xmm6);
+  __ mulsd(xmm1, xmm4);
+  __ movsd(xmm4, ExternalAddress(ONEMASK), r11 /*rscratch*/);
+  __ subsd(xmm3, xmm6);
+  __ xorpd(xmm0, xmm0);
+  __ addsd(xmm2, xmm1);
+  __ subsd(xmm7, xmm3);
+  __ movdqu(xmm6, ExternalAddress(cv), r11 /*rscratch*/);
+  __ addsd(xmm2, xmm7);
+  __ movl(rcx, 255);
+  __ andl(rcx, rax);
+  __ addl(rcx, rcx);
+  __ lea(r8, ExternalAddress(T2_neg_f));
+  __ movdqu(xmm5, Address(r8, rcx, Address::times(8)));
+  __ shrl(rax, 4);
+  __ andl(rax, 65520);
+  __ subl(rax, 16368);
+  __ negl(rax);
+  __ pinsrw(xmm0, rax, 3);
+  __ movdqu(xmm1, ExternalAddress(cv + 16), r11 /*rscratch*/);
+  __ pshufd(xmm0, xmm0, 68);
+  __ mulpd(xmm0, xmm5);
+  __ movsd(xmm7, ExternalAddress(cv + 32), r11 /*rscratch*/);
+  __ pshufd(xmm2, xmm2, 68);
+  __ movq(xmm5, xmm4);
+  __ addsd(xmm4, xmm0);
+  __ mulpd(xmm6, xmm2);
+  __ mulsd(xmm7, xmm2);
+  __ mulpd(xmm2, xmm2);
+  __ addpd(xmm1, xmm6);
+  __ mulsd(xmm2, xmm2);
+  __ movsd(xmm3, ExternalAddress(ONEMASK), r11 /*rscratch*/);
+  __ mulpd(xmm1, xmm2);
+  __ pshufd(xmm6, xmm1, 78);
+  __ addsd(xmm1, xmm6);
+  __ movq(xmm6, xmm1);
+  __ addsd(xmm1, xmm7);
+  __ mulsd(xmm1, xmm0);
+  __ addsd(xmm1, xmm4);
+  __ andpd(xmm4, ExternalAddress(MASK3), r11 /*rscratch*/);
+  __ divsd(xmm5, xmm1);
+  __ subsd(xmm3, xmm4);
+  __ pshufd(xmm1, xmm0, 238);
+  __ addsd(xmm3, xmm0);
+  __ movq(xmm2, xmm4);
+  __ addsd(xmm3, xmm1);
+  __ mulsd(xmm1, xmm7);
+  __ mulsd(xmm7, xmm0);
+  __ addsd(xmm3, xmm1);
+  __ addsd(xmm4, xmm7);
+  __ movsd(xmm1, ExternalAddress(RMASK), r11 /*rscratch*/);
+  __ mulsd(xmm6, xmm0);
+  __ andpd(xmm4, ExternalAddress(MASK3), r11 /*rscratch*/);
+  __ addsd(xmm3, xmm6);
+  __ movq(xmm6, xmm4);
+  __ subsd(xmm2, xmm4);
+  __ addsd(xmm2, xmm7);
+  __ movsd(xmm7, ExternalAddress(ONEMASK), r11 /*rscratch*/);
+  __ andpd(xmm5, xmm1);
+  __ addsd(xmm3, xmm2);
+  __ mulsd(xmm4, xmm5);
+  __ xorpd(xmm2, xmm2);
+  __ mulsd(xmm3, xmm5);
+  __ subsd(xmm6, ExternalAddress(TWOMASK), r11 /*rscratch*/);
+  __ subsd(xmm4, xmm7);
+  __ xorl(rdx, 32768);
+  __ pinsrw(xmm2, rdx, 3);
+  __ addsd(xmm4, xmm3);
+  __ mulsd(xmm6, xmm5);
+  __ movq(xmm1, xmm3);
+  __ mulsd(xmm3, xmm4);
+  __ movq(xmm0, xmm6);
+  __ mulsd(xmm6, xmm4);
+  __ subsd(xmm1, xmm3);
+  __ subsd(xmm1, xmm6);
+  __ addsd(xmm0, xmm1);
+  __ xorpd(xmm0, xmm2);
+  __ jmp(B1_4);
+
+  __ bind(L_2TAG_PACKET_0_0_1);
+  __ addl(rcx, 960);
+  __ cmpl(rcx, 1104);
+  __ jcc(Assembler::aboveEqual, L_2TAG_PACKET_1_0_1);
+  __ movdqu(xmm2, ExternalAddress(pv), r11 /*rscratch*/);
+  __ pshufd(xmm1, xmm0, 68);
+  __ movdqu(xmm3, ExternalAddress(pv + 16), r11 /*rscratch*/);
+  __ mulpd(xmm1, xmm1);
+  __ movdqu(xmm4, ExternalAddress(pv + 32), r11 /*rscratch*/);
+  __ mulpd(xmm2, xmm1);
```

Comment:
I would encourage either you add detailed comments or give meaningful names to the registers to ease the review process.

Reply:
I agree, this is all rather obscure. Ideally the same names that are used in wherever this comes from.

Where does the algorithm come from? What are its accuracy guarantees?

In addition, given the rarity of hyperbolic tangents in Java applications, do we need this?

Reply:
@theRealAph, this implementation is based on Intel libm math library and meets the accuracy requirements. The algorithm is provided in the comments.

Reply:
Do you have a copy of this information? Should it be in the commit?

Reply:
@vamsi-parasa  don't hesitate in adding as much and explicit information about the original source from where the algorithm has been picked up, even though the PR explicitly mentions libm. Adding the link to source references is a good practice.

Reply:
@jatin-bhateja This is based on Intel internal LIBM sources and so there is no public link available.

Reply:
> Do you have a copy of this information? Should it be in the commit?

@theRealAph The accuracy of standard (non fast mode) LIBM functions ensures errors of < 1 ulp. LIBM is part of Intel C++ compiler. The documentation can be found here: https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/programming-tradeoffs-floating-point-applications.html.